File: server/src/models/tag.py
from datetime import datetime, timezone
from enum import Enum
from uuid import uuid4
from pydantic import BaseModel, Field

class DataType(str, Enum):
    FLOAT = "float"
    INT = "int"
    BOOL = "bool"
    STRING = "string"

class Criticality(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"

class TagStatus(str, Enum):
    ACTIVE = "active"
    RETIRED = "retired"
    DRAFT = "draft"

class ApprovalStatus(str, Enum):
    NONE = "none"
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"

class Tag(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    name: str
    description: str
    unit: str                  # e.g., "bar", "°C", "RPM"
    datatype: DataType
    samplingFrequency: float   # seconds
    criticality: Criticality
    status: TagStatus = TagStatus.DRAFT
    approvalStatus: ApprovalStatus = ApprovalStatus.NONE
    rejectionReason: str | None = None
    assetId: str               # FK to Asset
    sourceId: str | None = None  # FK to Source, optional
    createdAt: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updatedAt: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class CreateTag(BaseModel):
    """Request body for POST /api/tags."""
    name: str
    description: str
    unit: str
    datatype: DataType
    samplingFrequency: float
    criticality: Criticality
    assetId: str
    sourceId: str | None = None


class UpdateTag(BaseModel):
    """Request body for PUT /api/tags/{id}. All fields optional for partial updates."""
    name: str | None = None
    description: str | None = None
    unit: str | None = None
    datatype: DataType | None = None
    samplingFrequency: float | None = None
    criticality: Criticality | None = None
    status: TagStatus | None = None
    sourceId: str | None = None
    assetId: str | None = None

File: client/src/types/tag.ts
/**
 * Tag-related type definitions.
 * Mirrors server/src/models/tag.py
 */

export type DataType = "float" | "int" | "bool" | "string";

export type Criticality = "low" | "medium" | "high" | "critical";

export type TagStatus = "active" | "retired" | "draft";

export type ApprovalStatus = "none" | "pending" | "approved" | "rejected";

export interface Tag {
  id: string;
  name: string;
  description: string;
  unit: string;
  datatype: DataType;
  samplingFrequency: number;
  criticality: Criticality;
  status: TagStatus;
  assetId: string;
  sourceId: string | null;
  approvalStatus: ApprovalStatus;
  rejectionReason: string | null;
  createdAt: string;
  updatedAt: string;
}

/** Request body for POST /api/tags */
export interface CreateTag {
  name: string;
  description: string;
  unit: string;
  datatype: DataType;
  samplingFrequency: number;
  criticality: Criticality;
  assetId: string;
  sourceId?: string | null;
}

/** Request body for PUT /api/tags/:id — all fields optional for partial updates */
export interface UpdateTag {
  name?: string;
  description?: string;
  unit?: string;
  datatype?: DataType;
  samplingFrequency?: number;
  criticality?: Criticality;
  status?: TagStatus;
  sourceId?: string | null;
}

/** A single validation error from the naming validator */
export interface NameValidationError {
  segment: string;
  message: string;
  received: string;
  expected: string | null;
}

/** Response from POST /api/tags/validate-name */
export interface ValidateNameResponse {
  valid: boolean;
  errors: NameValidationError[];
}

/** Request body for POST /api/tags/auto-fill */
export interface AutoFillRequest {
  query: string;
}

/** A single search hit from the golden-tags index */
export interface AutoFillMatch {
  tagName: string;
  description: string;
  score: number;
  site: string;
  line: string;
  equipment: string;
  unit: string;
  datatype: string;
}

/** Response from POST /api/tags/auto-fill */
export interface AutoFillResult {
  site: string | null;
  line: string | null;
  equipment: string | null;
  unit: string | null;
  datatype: string | null;
  name: string | null;
  description: string | null;
  criticality: string | null;
  confidence: number;
  matches: AutoFillMatch[];
}
